File: packages/core/src/schemas.ts
import { z } from 'zod';

// Battle schemas
export const BattleStatusSchema = z.enum(['scheduled', 'open', 'closed']);
export type BattleStatus = z.infer<typeof BattleStatusSchema>;

export const BattleSchema = z.object({
  id: z.string().uuid(),
  event_id: z.string().uuid(),
  mc_a: z.string(),
  mc_b: z.string(),
  starts_at: z.string().datetime(),
  ends_at: z.string().datetime(),
  status: BattleStatusSchema,
});
export type Battle = z.infer<typeof BattleSchema>;

// Vote schemas
export const VoteChoiceSchema = z.enum(['A', 'B']);
export type VoteChoice = z.infer<typeof VoteChoiceSchema>;

export const VoteRequestSchema = z.object({
  battle_id: z.string().uuid(),
  choice: VoteChoiceSchema,
  device_hash: z.string().min(1),
});
export type VoteRequest = z.infer<typeof VoteRequestSchema>;

export const VoteResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  tally: z.object({
    A: z.number(),
    B: z.number(),
  }).optional(),
});
export type VoteResponse = z.infer<typeof VoteResponseSchema>;

// Tally schemas
export const TallySchema = z.object({
  A: z.number(),
  B: z.number(),
});
export type Tally = z.infer<typeof TallySchema>;

// Admin schemas
export const AdminOpenBattleSchema = z.object({
  starts_at: z.string().datetime().optional(),
  ends_at: z.string().datetime().optional(),
});
export type AdminOpenBattle = z.infer<typeof AdminOpenBattleSchema>;

// Error schemas
export const ErrorResponseSchema = z.object({
  detail: z.string(),
});
export type ErrorResponse = z.infer<typeof ErrorResponseSchema>;

File: apps/api/schemas.py
"""Pydantic schemas for API requests and responses."""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from models import BattleStatus, VoteChoice


class BattleResponse(BaseModel):
    """Battle response schema."""
    id: uuid.UUID
    event_id: uuid.UUID
    mc_a: str
    mc_b: str
    starts_at: datetime
    ends_at: datetime
    status: BattleStatus

    class Config:
        from_attributes = True


class VoteRequest(BaseModel):
    """Vote request schema."""
    battle_id: uuid.UUID
    choice: VoteChoice
    device_hash: str = Field(..., min_length=1)


class TallyResponse(BaseModel):
    """Tally response schema."""
    A: int
    B: int


class VoteResponse(BaseModel):
    """Vote response schema."""
    success: bool
    message: str
    tally: Optional[TallyResponse] = None


class AdminCreateBattleRequest(BaseModel):
    """Admin create battle request schema."""
    event_id: str
    mc_a: str
    mc_b: str
    starts_at: datetime
    ends_at: datetime


class AdminOpenBattleRequest(BaseModel):
    """Admin open battle request schema."""
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None


class HealthResponse(BaseModel):
    """Health check response schema."""
    status: str
